x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/esql/EsqlDataExtractor.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.datafeed.extractor.esql;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.json.JsonXContent;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfig;
+import org.elasticsearch.xpack.core.ml.datafeed.SearchInterval;
+import org.elasticsearch.xpack.esql.action.ColumnInfo;
+import org.elasticsearch.xpack.esql.action.EsqlQueryAction;
+import org.elasticsearch.xpack.esql.action.EsqlQueryRequest;
+import org.elasticsearch.xpack.esql.action.EsqlQueryResponse;
+import org.elasticsearch.xpack.ml.datafeed.extractor.DataExtractor;
+import org.elasticsearch.xpack.ql.util.DateUtils;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class EsqlDataExtractor implements DataExtractor {
+
+    private static final Logger logger = LogManager.getLogger(EsqlDataExtractor.class);
+
+    private final Client client;
+    private final DatafeedConfig datafeed;
+    private final String timeField;
+    private final SearchInterval interval;
+    private boolean isCancelled;
+
+    EsqlDataExtractor(Client client, DatafeedConfig datafeed, String timeField, long start, long end) {
+        this.client = Objects.requireNonNull(client);
+        this.datafeed = Objects.requireNonNull(datafeed);
+        this.timeField = timeField;
+        this.interval = new SearchInterval(start, end);
+        this.isCancelled = false;
+    }
+
+    // TODO: check whether these expressions facilitate injection attacks!
+    private String esqlTimeFilter() {
+        return Strings.format(
+            " | WHERE %s >= \"%s\" AND %s < \"%s\"",
+            timeField,
+            DateUtils.UTC_DATE_TIME_FORMATTER.formatMillis(
+                Math.min(interval.startMs(), org.elasticsearch.common.time.DateUtils.MAX_MILLIS_BEFORE_9999)
+            ),
+            timeField,
+            DateUtils.UTC_DATE_TIME_FORMATTER.formatMillis(
+                Math.min(interval.endMs(), org.elasticsearch.common.time.DateUtils.MAX_MILLIS_BEFORE_9999)
+            )
+        );
+    }
+
+    private String esqlSortByTime() {
+        return Strings.format(" | SORT %s", timeField);
```

Comment:
On the theme of other comments, this needs to be the time field that's present at the end of the pipeline.

In the case where the pipeline is aggregating data I can see that being different to the time field at the start. For example, if the user is using ES|QL to bucket data to 10 minute chunks then maybe the input time field would be `@timestamp` and the output time field for section of ES|QL that they supplied would be `bucket_timestamp`.

Reply:
I think it's okay to enforce that the "input time field name" equals the "output time field name".

That's pretty similar to a [requirement in the query DSL datafeed](https://www.elastic.co/guide/en/machine-learning/current/ml-configuring-aggregation.html#aggs-using-date-histogram):
> The aggregations have names that match the fields that they operate on